zeitgeistpm/zeitgeist
# [CICD] Add benchmark workflow

**Issue**
#109 

**Solution**
A workflow is created that contains a bash script that executes benchmarks.

1. Watch for pushes on a *stage-benchmark* branch
2. Clone the latest state of the *stage-benchmark* branch
3. Run (optimized) runtime-benchmarks tests
4. Build optimized zeitgeist binary with runtime-benchmarks features
5. Run our benchmark script - For every benchmark command it:
   a.  Checks if the *stage-benchmark* branch contains new commits and if yes, aborts the benchmarks and does NOT commit changes
   b. Runs the benchmark
6. Checks the branch again for new commits and if none are existent, pushes the updated weights to the *stage-benchmark* branch

We can also run this script directly on the main branch, but I suggest that we use a staging branch to:
1. Verify that everything works appropriately (it was tested already, but who knows)
2. Squash the two commits, the commit that contains the PR merge and the "update weights" commit, into one commit and merge only one commit into the main branch. If we don't do this, every or almost every second commit on the main branch will be "Update weights".

This workflow can be used now, but I suggest to not use it due to the low hardware specs of the default runner and the uncertainty regarding the availability of the hardware resources. More info on #187 and #185. In addition, creating an action that is ready for the marketplace and public usage is also a nice-to-have #184.

You can inspect a [successfully executed benchmark workflow](https://github.com/zeitgeistpm/zeitgeist/runs/2848623725?check_suite_focus=true) in the past runs of this repository and its corresponding automatically committed new weights at commit hash [c6754e2cc608a2301d1f956f6eac8a4fe94f9a2e](https://github.com/zeitgeistpm/zeitgeist/pull/181/commits/c6754e2cc608a2301d1f956f6eac8a4fe94f9a2e)
